api/v1alpha1/owner_labels.go

```diff
@@ -0,0 +1,229 @@
+	})
+}
+
+// AddPersistentStorageLabels adds labels to a resource to indicate which persistent storage instance it belongs to
+func AddPersistentStorageLabels(child metav1.Object, persistentStorage *PersistentStorageInstance) {
+	labels := child.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels[PersistentStorageNameLabel] = persistentStorage.Name
+	labels[PersistentStorageNamespaceLabel] = persistentStorage.Namespace
+
+	child.SetLabels(labels)
+}
+
+// MatchingPersistentStorage returns the MatchingLabels to match the persistent storage labels
+func MatchingPersistentStorage(persistentStorage *PersistentStorageInstance) client.MatchingLabels {
+	return client.MatchingLabels(map[string]string{
+		PersistentStorageNameLabel:      persistentStorage.Name,
+		PersistentStorageNamespaceLabel: persistentStorage.Namespace,
+	})
+}
+
+// InheritParentLabels adds all labels from a parent resource to a child resource, excluding
+// the owner labels
+func InheritParentLabels(child metav1.Object, owner metav1.Object) {
+	labels := child.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	for key, value := range owner.GetLabels() {
+		// don't inherit owner labels
+		if key == OwnerNameLabel || key == OwnerNamespaceLabel || key == OwnerKindLabel {
+			continue
+		}
+
+		labels[key] = value
+	}
+
+	child.SetLabels(labels)
+}
+
+// deleteChildrenSingle deletes all the children of a single type. Children are found using
+// the owner labels
+func deleteChildrenSingle(ctx context.Context, c client.Client, childObjectList ObjectList, parent metav1.Object) (DeleteStatus, error) {
+	// List all the children and filter by the owner labels
+	err := c.List(ctx, childObjectList.(client.ObjectList), MatchingOwner(parent))
+	if err != nil {
+		return DeleteRetry, err
+	}
+
+	objectList := childObjectList.GetObjectList()
+	if len(objectList) == 0 {
+		return DeleteComplete, nil
+	}
+
+	// Check whether the child objects span multiple namespaces.
+	multipleNamespaces := false
+	namespace := ""
+	for _, obj := range objectList {
+		if obj.GetNamespace() != namespace && namespace != "" {
+			multipleNamespaces = true
+		}
+
+		namespace = obj.GetNamespace()
+
+		// Wait for any deletes to finish if the resource is already
+		// marked for deletion
+		if !obj.GetDeletionTimestamp().IsZero() {
+			return DeleteRetry, nil
+		}
+	}
+
+	// If all the child resources are in a single namespace then we can use DeleteAllOf
+	if !multipleNamespaces {
+		err = c.DeleteAllOf(ctx, objectList[0], client.InNamespace(parent.GetNamespace()), MatchingOwner(parent))
+		if err != nil {
+			return DeleteRetry, err
+		}
+
+		return DeleteRetry, nil
+	}
+
+	// If the child resources span multiple namespaces, then we have to delete them
+	// each individually.
+	g := new(errgroup.Group)
+	for _, obj := range objectList {
+		obj := obj
+
+		// Start a goroutine for each object to delete
+		g.Go(func() error {
```

Comment:
You're tossing these onto a queue of go routines, and that could mean a delay before the first one runs and an even   longer delay before the last one runs.  These could get stuck behind a bunch of reconciles for other Workflows.

Reply:
I ran some tests a few weeks back creating and deleting 10,000 resources. It ended up taking 7 or 8 minutes to delete them when doing it serially because of the latency with making the API server request. Even if these go routines have to wait to get scheduled it should still be way faster than deleting them serially.